radixdlt-engine/src/main/java/com/radixdlt/application/tokens/construction/UnstakeOwnershipConstructor.java

```diff
@@ -19,22 +19,45 @@
 package com.radixdlt.application.tokens.construction;
 
 import com.radixdlt.atom.ActionConstructor;
+import com.radixdlt.atom.SubstateTypeId;
 import com.radixdlt.atom.TxBuilder;
 import com.radixdlt.atom.TxBuilderException;
 import com.radixdlt.atom.actions.UnstakeOwnership;
 import com.radixdlt.application.system.state.StakeOwnership;
 import com.radixdlt.application.tokens.state.PreparedUnstakeOwnership;
+import com.radixdlt.constraintmachine.SubstateIndex;
+import com.radixdlt.crypto.ECPublicKey;
+
+import java.nio.ByteBuffer;
 
 public class UnstakeOwnershipConstructor implements ActionConstructor<UnstakeOwnership> {
 	@Override
 	public void construct(UnstakeOwnership action, TxBuilder txBuilder) throws TxBuilderException {
-		txBuilder.swapFungible(
-			StakeOwnership.class,
+		if (action.amount().isZero()) {
+			throw new TxBuilderException("Must transfer > 0.");
+		}
+
+		var buf = ByteBuffer.allocate(2 + ECPublicKey.COMPRESSED_BYTES + (1 + ECPublicKey.COMPRESSED_BYTES));
+		buf.put(SubstateTypeId.STAKE_OWNERSHIP.id());
+		buf.put((byte) 0);
+		buf.put(action.from().getCompressedBytes());
+		buf.put(action.accountAddr().getBytes());
```

Comment:
I see several places where this (or similar, depending on the action) logic is repeated. Perhaps it worth to create dedicated class to handle all these calculations/serializations?